File: internal/config/db/psql/connection.py
from sqlalchemy import create_engine
import os


def make_connection_string_from_env() -> str:
    pas = os.getenv("POSTGRES_PASSWORD")
    user = os.getenv("POSTGRES_USER")
    host = os.getenv("POSTGRES_HOST")
    port = os.getenv("POSTGRES_PORT")
    db = os.getenv("POSTGRES_DB")

    s = f"postgresql://{user}:{pas}@{host}:{port}/{db}"
    return s
    

dsn = make_connection_string_from_env()

engine = create_engine(dsn)

class DB:
    def __init__(self):
        self.engine = engine
    
    def get_engine(self):
        return self.engine
    def close(self):
        self.engine.dispose()

File: service/pkg/database/psql/connection.go
package psql

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"github.com/jmoiron/sqlx"
	"github.com/joho/godotenv"
	_ "github.com/lib/pq"
)

var DB *sqlx.DB

func Init() {
	if err := godotenv.Load(); err != nil {
		log.Printf("no .env file loaded: %v", err)
	}

	ctx := context.Background()

	databaseURL := getEnv("DATABASE_URL", "")

	host := getEnv("POSTGRES_HOST", "")
	port := getEnv("POSTGRES_PORT", "")
	user := getEnv("POSTGRES_USER", "")
	password := getEnv("POSTGRES_PASSWORD", "")
	dbname := getEnv("POSTGRES_DB", "")
	sslmode := getEnv("POSTGRES_SSLMODE", "")

	var dsn string
	if databaseURL != "" {
		dsn = databaseURL
	} else {
		dsn = fmt.Sprintf("user=%s password=%s dbname=%s host=%s port=%s sslmode=%s", user, password, dbname, host, port, sslmode)
	}

	db, err := sqlx.ConnectContext(ctx, "postgres", dsn)
	if err != nil {
		log.Printf("Failed to connect to database: %v", err)
	}
	if err := db.Ping(); err != nil {
		log.Printf("Failed to ping database: %v", err)
	}
	log.Printf("Connected to database successfully")

	DB = db

	DB.SetMaxOpenConns(5)
	DB.SetMaxIdleConns(2)
	DB.SetConnMaxLifetime(60 * time.Minute)
	DB.SetConnMaxIdleTime(30 * time.Minute)
}

func getEnv(key, fallback string) string {
	if val, ok := os.LookupEnv(key); ok && val != "" {
		return val
	}
	return fallback
}
func GetDB() *sqlx.DB {
	return DB
}
